processing/src/main/java/io/druid/segment/incremental/IncrementalIndex.java

```diff
@@ -1304,4 +1293,240 @@ public int compare(TimeAndDims lhs, TimeAndDims rhs)
+      if (sortFacts) {
+        return ((ConcurrentNavigableMap<TimeAndDims, Integer>) facts).lastKey().getTimestamp();
+      } else {
+        throw new UnsupportedOperationException("can't get maxTime from unsorted facts data.");
+      }
+    }
+
+    @Override
+    public Iterator<Pair<TimeAndDims, Integer>> iterator()
+    {
+      return iterator(false);
+    }
+
+    @Override
+    public Iterator<Pair<TimeAndDims, Integer>> iterator(boolean descending)
+    {
+      if (descending && sortFacts) {
+        return wrap(((ConcurrentNavigableMap<TimeAndDims, Integer>) facts).descendingMap().entrySet());
+      }
+      return wrap(facts.entrySet());
+    }
+
+    @Override
+    public Iterable<Pair<TimeAndDims, Integer>> timeRangeIterable(boolean descending, long timeStart, long timeEnd)
+    {
+      if (!sortFacts) {
+        throw new UnsupportedOperationException("can't get timeRange from unsorted facts data.");
+      }
+      TimeAndDims start = new TimeAndDims(timeStart, new int[][]{});
+      TimeAndDims end = new TimeAndDims(timeEnd, new int[][]{});
+      ConcurrentNavigableMap<TimeAndDims, Integer> subMap =
+          ((ConcurrentNavigableMap<TimeAndDims, Integer>) facts).subMap(start, end);
+      final Map<TimeAndDims, Integer> rangeMap = descending ? subMap.descendingMap() : subMap;
+      return new Iterable<Pair<TimeAndDims, Integer>>()
+      {
+        @Override
+        public Iterator<Pair<TimeAndDims, Integer>> iterator()
+        {
+          return wrap(rangeMap.entrySet());
+        }
+      };
+    }
+
+    private Iterator<Pair<TimeAndDims, Integer>> wrap(Iterable<Map.Entry<TimeAndDims, Integer>> iterable)
+    {
+      return Iterators.transform(
+          iterable.iterator(),
+          new Function<Map.Entry<TimeAndDims, Integer>, Pair<TimeAndDims, Integer>>()
+          {
+            @Override
+            public Pair<TimeAndDims, Integer> apply(Map.Entry<TimeAndDims, Integer> input)
+            {
+              return new Pair<>(input.getKey(), input.getValue());
+            }
+          }
+      );
+    }
+
+    @Override
+    public Integer putIfAbsent(TimeAndDims key, Integer rowIndex)
+    {
+      return facts.putIfAbsent(key, rowIndex);
+    }
+
+    @Override
+    public void clear()
+    {
+      facts.clear();
+    }
+  }
+
+  static class PlainFactsHolder implements FactsHolder
+  {
+    private final boolean sortFacts;
+    private final ConcurrentMap<Long, Queue<Pair<TimeAndDims, Integer>>> facts;
+
+    public PlainFactsHolder(boolean sortFacts)
+    {
+      this.sortFacts = sortFacts;
+      if (sortFacts) {
+        this.facts = new ConcurrentSkipListMap<>(new Comparator<Long>()
+        {
+          @Override
+          public int compare(Long lhs, Long rhs)
+          {
+            return Longs.compare(lhs, rhs);
+          }
+        });
+      } else {
+        this.facts = new ConcurrentHashMap<>();
+      }
+    }
+
+    @Override
+    public Integer getPriorIndex(TimeAndDims key)
```

Comment:
How about renaming this method to getRowIndexForRollup(TimeAndDims key) to make it more clear that the underlying impl may not always be a map.